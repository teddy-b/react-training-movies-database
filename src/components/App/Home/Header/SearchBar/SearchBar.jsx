import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './SearchBar.scss'
import Logo from '../../../../shared/Logo'
import { SEARCH_BY, GENRES } from '../../../../../constants/global'

class SearchBar extends Component {
  constructor(props) {
    super(props)
    this.inputRef = React.createRef()
  }

  render() {
    const {
      onSearch,
      onSearchMoviesByGenre,
      onSearchMoviesByTitle,
      searchBy
    } = this.props

    return (
      <div className="searchBar">
        <Logo />
        <p className="searchBarTitle">Find your movie</p>
        <input
          placeholder="Start typing movie title or genre"
          className="searchInput"
          type="text"
          ref={this.inputRef}
          onKeyPress={event => event.key === 'Enter' && onSearch(event.target.value, searchBy)}
          list="genres"
        />
        {searchBy === SEARCH_BY.genre &&
          <datalist id="genres">
            {GENRES.map(genre => <option key={genre} value={genre} />)}
          </datalist>
        }
        <div className="searchFooter">
          <div className="searchBy">Search by </div>
          <div className="searchByBtns">
            <button
              className={`searchByTitleBtn ${searchBy === SEARCH_BY.title && 'selected'}`}
              onClick={onSearchMoviesByTitle}
            >
              Title
            </button>
            <button
              className={`searchByGenreBtn ${searchBy === SEARCH_BY.genre && 'selected'}`}
              onClick={onSearchMoviesByGenre}
            >
              Genre
            </button>
          </div>
          <button
            className="searchBtn"
            onClick={() => onSearch(this.inputRef.current.value, searchBy)}
          >
            Search
          </button>
        </div>
      </div>
    )
  }
}

SearchBar.propTypes = {
  onSearch: PropTypes.func.isRequired,
  onSearchMoviesByGenre: PropTypes.func.isRequired,
  onSearchMoviesByTitle: PropTypes.func.isRequired,
  searchBy: PropTypes.string.isRequired
}

export default SearchBar
