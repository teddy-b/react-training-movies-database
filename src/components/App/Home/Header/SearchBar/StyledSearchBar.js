import styled from 'styled-components'

import { black, grey2, grey4, red, white } from '../../../../../constants/styles'

export const StyledSearchBar = styled.div`
  background-color: ${grey4};
  color: ${white};
  padding: .5em;
`

export const StyledSearchBarTitle = styled.p`
  text-transform: uppercase;
`

export const StyledSearchInput = styled.input`
  width: 98%;
  padding: .8em;
  background-color: ${black};
  color: inherit;
  border: none;
  border-bottom: 1px solid ${red};
`

export const StyledSearchFooter = styled.div`
  padding: .5em;
  display: flex;
  justify-content: space-between;
  align-items: center;
`

export const StyledSearchBy = styled.div`
  text-transform: uppercase;
`

export const StyledSearchByBtns = styled.div`
  flex-grow: 2;
`

export const StyledSearchByBtn = styled.button`
  color: ${white};
  border: none;
  cursor: pointer;
  margin: .3em;
  padding: .6em;
  background-color: ${props => (props.selected ? red : grey2)};
`

export const StyledSearchBtn = styled.button`
  color: ${white};
  border: none;
  cursor: pointer;
  margin: .2em;
  padding: .8em;
  width: 15vw;
  background-color: ${red};
`
