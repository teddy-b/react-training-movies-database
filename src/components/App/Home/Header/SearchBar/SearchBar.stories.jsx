/* @flow */

import * as React from 'react'
import { MemoryRouter } from 'react-router-dom'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import type { Renderable } from '@storybook/react'

import SearchBar from './SearchBar'
import { SEARCH_BY } from '../../../../../constants/global'

const props = {
  onSearch: action('clicked'),
  onSearchMoviesByGenre: action('clicked'),
  onSearchMoviesByTitle: action('clicked'),
  searchBy: SEARCH_BY.title
}

storiesOf('Components/SearchBar', module)
  .add('search by title', (): Renderable => (
    <MemoryRouter>
      <SearchBar
        {...props}
        searchBy={SEARCH_BY.title}
      />
    </MemoryRouter>
  ))
  .add('search by genre', (): Renderable => (
    <MemoryRouter>
      <SearchBar
        {...props}
        searchBy={SEARCH_BY.genre}
      />
    </MemoryRouter>
  ))
