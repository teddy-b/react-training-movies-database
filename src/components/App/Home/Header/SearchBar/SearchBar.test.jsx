import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'

import SearchBar from './SearchBar'

jest.mock('../../../../shared/Logo', () => 'data-logo')

describe('SearchBar', () => {
  const props = {
    onSearch: jest.fn(),
    onSearchMoviesByGenre: jest.fn(),
    onSearchMoviesByTitle: jest.fn(),
    searchBy: 'title'
  }

  it('renders correctly with search by title', () => {
    const component = renderer.create(<SearchBar {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with search by genres', () => {
    const component = renderer.create(
      <SearchBar
        {...props}
        searchBy="genres"
      />
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should trigger onSearchMoviesByTitle onClick', () => {
    const onSearchMoviesByTitleMock = jest.fn()
    const component = shallow(
      <SearchBar
        {...props}
        onSearchMoviesByTitle={onSearchMoviesByTitleMock}
        searchBy="genres"
      />
    )

    component.find('button').at(0).simulate('click')

    expect(onSearchMoviesByTitleMock).toBeCalled()
  })

  it('should trigger onSearchMoviesByGenre onClick', () => {
    const onSearchMoviesByGenreMock = jest.fn()
    const component = shallow(
      <SearchBar
        {...props}
        onSearchMoviesByGenre={onSearchMoviesByGenreMock}
      />
    )

    component.find('button').at(1).simulate('click')

    expect(onSearchMoviesByGenreMock).toBeCalled()
  })

  it('should trigger onSearch onClick', () => {
    const onSearchMock = jest.fn()
    const component = mount(
      <SearchBar
        {...props}
        onSearch={onSearchMock}
      />
    )

    component.find('input').at(0).simulate('change', { target: { value: 'New value' } })
    component.find('button').at(2).simulate('click')

    expect(onSearchMock).toBeCalled()
  })

  it('should trigger onSearch onKeyPress', () => {
    const onSearchMock = jest.fn()
    const component = mount(
      <SearchBar
        {...props}
        onSearch={onSearchMock}
        searchBy="title"
      />
    )

    component.find('input').at(0)
      .simulate('change', { target: { value: 'New value' } })
      .simulate('keypress', { key: 'Enter' })

    expect(onSearchMock).toBeCalled()
  })
})
