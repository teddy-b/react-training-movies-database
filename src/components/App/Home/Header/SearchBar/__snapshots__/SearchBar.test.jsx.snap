// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders correctly with search by genres 1`] = `
<div
  className="searchBar"
>
  <data-logo />
  <p
    className="searchBarTitle"
  >
    Find your movie
  </p>
  <input
    className="searchInput"
    list="genres"
    onKeyPress={[Function]}
    placeholder="Start typing movie title or genre"
    type="text"
  />
  <datalist
    id="genres"
  >
    <option
      value="Action"
    />
    <option
      value="Adventure"
    />
    <option
      value="Animation"
    />
    <option
      value="Comedy"
    />
    <option
      value="Crime"
    />
    <option
      value="Documentary"
    />
    <option
      value="Drama"
    />
    <option
      value="Family"
    />
    <option
      value="Fantasy"
    />
    <option
      value="History"
    />
    <option
      value="Horror"
    />
    <option
      value="Music"
    />
    <option
      value="Mystery"
    />
    <option
      value="Romance"
    />
    <option
      value="Science Fiction"
    />
    <option
      value="Thriller"
    />
    <option
      value="TV Movie"
    />
    <option
      value="War"
    />
    <option
      value="Western"
    />
  </datalist>
  <div
    className="searchFooter"
  >
    <div
      className="searchBy"
    >
      Search by 
    </div>
    <div
      className="searchByBtns"
    >
      <button
        className="searchByTitleBtn false"
        onClick={[MockFunction]}
      >
        Title
      </button>
      <button
        className="searchByGenreBtn selected"
        onClick={[MockFunction]}
      >
        Genre
      </button>
    </div>
    <button
      className="searchBtn"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;

exports[`SearchBar renders correctly with search by title 1`] = `
<div
  className="searchBar"
>
  <data-logo />
  <p
    className="searchBarTitle"
  >
    Find your movie
  </p>
  <input
    className="searchInput"
    list="genres"
    onKeyPress={[Function]}
    placeholder="Start typing movie title or genre"
    type="text"
  />
  <div
    className="searchFooter"
  >
    <div
      className="searchBy"
    >
      Search by 
    </div>
    <div
      className="searchByBtns"
    >
      <button
        className="searchByTitleBtn selected"
        onClick={[MockFunction]}
      >
        Title
      </button>
      <button
        className="searchByGenreBtn false"
        onClick={[MockFunction]}
      >
        Genre
      </button>
    </div>
    <button
      className="searchBtn"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;
